syntax = "proto3";

package tinyrefl.tool.model;

message SourceLocation
{
    optional string file   = 1;
    optional uint64 line   = 2;
    optional uint64 column = 3;
}

message SourceRange
{
    optional SourceLocation begin = 1;
    optional SourceLocation end   = 2;
}

message Identifier
{
    optional string name            = 1;
    optional string fullName        = 2;
    optional string displayName     = 3;
    optional string fullDisplayName = 4;
    optional uint64 uniqueId        = 5;
}

message Declaration
{
    optional uint64         parentUniqueId  = 1;
    optional Identifier     id              = 2;
    optional SourceLocation sourceLocation  = 3;
    optional SourceRange    sourceRange     = 4;
    optional string         declarationText = 5;
}

message Type
{
    optional Declaration declaration = 1;
    optional Type        decayedType = 2;
}

message Attribute
{
    optional string name             = 1;
    optional string fullName         = 2;
    optional        string namespace = 3;
    optional string fullAttribute    = 4;
    repeated string arguments        = 5;
    optional string joinedArguments  = 6;
    optional uint64 uniqueId         = 7;
}

message Parameter
{
    optional string    name       = 1;
    optional Type      type       = 2;
    optional uint64    position   = 3;
    repeated Attribute attributes = 4;
}

message Function
{
    optional Declaration declaration = 1;
    repeated Parameter   parameters  = 2;
    optional Type        returnType  = 3;
    repeated Attribute   attributes  = 4;
}

message MemberFunction
{
    enum Kind
    {
        NonConst = 0;
        Const    = 1;
        Static   = 2;
    }

    optional Declaration declaration = 1;
    repeated Parameter   parameters  = 2;
    optional Type        returnType  = 3;
    repeated Attribute   attributes  = 4;
    optional Kind        kind        = 5;
    optional             Type class  = 6;
}

message Constructor
{
    optional Declaration declaration = 1;
    repeated Parameter   parameters  = 2;
    optional             Type class  = 3;
    repeated Attribute   attributes  = 4;
}

message Variable
{
    optional Declaration declaration = 1;
    optional Type        type        = 2;
    repeated Attribute   attributes  = 3;
}

message MemberVariable
{
    optional Declaration declaration = 1;
    optional Type        type        = 2;
    repeated Attribute   attributes  = 3;
    optional             Type class  = 4;
}

message EnumValue
{
    optional Declaration declaration = 1;
    optional uint64      index       = 2;
    optional string      value       = 3;
    repeated Attribute   attributes  = 4;
}

message Enum
{
    optional Declaration declaration = 1;
    optional Type        type        = 2;
    repeated EnumValue   values      = 3;
    repeated Attribute   attributes  = 4;
}

message ClassMember
{
    enum Access
    {
        Public    = 0;
        Protected = 1;
        Private   = 2;
    }

    Access access = 1;

    oneof member
    {
        MemberFunction function    = 2;
        MemberVariable variable    = 3;
        Constructor    constructor = 4;
        Class class                = 5;
        Enum enum                  = 6;
        Function staticFunction    = 7;
        Variable staticVariable    = 8;
    }
}

message Class
{
    optional Declaration declaration = 1;
    optional Type        type        = 2;
    repeated Type        baseClasses = 3;
    repeated ClassMember members     = 4;
    repeated Attribute   attributes  = 5;
}

message NamespaceMember
{
    oneof member
    {
        Namespace namespace = 2;
        Function function   = 3;
        Variable variable   = 4;
        Enum enum           = 5;
        Class class         = 6;
    }
}

message Namespace
{
    optional Declaration     declaration = 1;
    repeated NamespaceMember members     = 2;
    repeated Attribute       attributes  = 3;
}

message File
{
    optional Identifier id              = 1;
    optional Namespace  globalNamespace = 2;
}
